editor/gui/scene_tree_editor.cpp

```diff
@@ -226,46 +226,180 @@ void SceneTreeEditor::_add_nodes(Node *p_node, TreeItem *p_parent) {
 	// which the editor needs not to know about.
 
 	bool part_of_subscene = false;
+	HashMap<Node *, CachedNode>::Iterator I = node_cache.find(p_node);
 
 	if (!display_foreign && p_node->get_owner() != get_scene_node() && p_node != get_scene_node()) {
 		if ((show_enabled_subscene || can_open_instance) && p_node->get_owner() && (get_scene_node()->is_editable_instance(p_node->get_owner()))) {
 			part_of_subscene = true;
-			//allow
+			// allow.
 		} else {
+			if (I) {
+				// stale node, remove recursively.
+				_node_cache_remove(p_node, true);
+			}
 			return;
 		}
 	} else {
 		part_of_subscene = p_node != get_scene_node() && get_scene_node()->get_scene_inherited_state().is_valid() && get_scene_node()->get_scene_inherited_state()->find_node_by_path(get_scene_node()->get_path_to(p_node)) >= 0;
 	}
 
-	TreeItem *item = tree->create_item(p_parent);
+	TreeItem *item;
+
+	bool is_new = false;
+
+	if (I) {
+		item = I->value.item;
+	} else {
+		// We don't set the node_index cache here. As undo/redo might've just recreated this.
+		item = tree->create_item(p_parent);
+		CachedNode cached_node;
+		cached_node.item = item;
+		I = node_cache.insert(p_node, cached_node);
+		is_new = true;
+
+		// If we have children we might have just been added because of a change type operation.
+		// We need to re-parent all of our children.
+		int cc = p_node->get_child_count(false);
+		for (int i = 0; i < cc; i++) {
+			Node *c = p_node->get_child(i, false);
+			HashMap<Node *, CachedNode>::Iterator IC = node_cache.find(c);
+
+			if (IC) {
+				if (IC->value.item->get_parent() != item) {
+					TreeItem *p = IC->value.item->get_parent();
+					if (p) {
+						p->remove_child(IC->value.item);
+					}
+					item->add_child(IC->value.item);
+				}
+			}
+		}
+	}
+
+	bool dirty = force_update || I->value.dirty;
 
-	item->set_text(0, p_node->get_name());
-	item->set_text_overrun_behavior(0, TextServer::OVERRUN_NO_TRIMMING);
+	if (p_parent && dirty) {
+		bool reparent = false;
+		int current_node_index = p_node->get_index(false);
```

Comment:
SceneTree dock only shows non-internal nodes, so internal ones are irrelevant.

Reply:
I looked at this, and I skip internal children whenever it matters whether there's expected to be TreeItems associated with the Nodes. As far as I can see internal nodes never get TreeItems, so I skip them whenever it matters (like for the node re-ordering, I think it would break if there are internal nodes)